### PR TITLE
001 test view job roles - Automated UI tests

### PR DESCRIPTION
**Pull request by lexseaton**

### CODE REVIEW

**Review by dylanlaffin — APPROVED** [review]
array well built

File: test/ui/viewJobRoleTests.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        driver = new Builder()
+            .forBrowser('chrome')
+            .setChromeOptions(new chrome.Options())
+            .build();
+
+    });
+
+    after(async function () {
+        try {
+            await driver.quit(); // Ensure this is awaited
```
Comment (boydstepkainos) [inline test/ui/viewJobRoleTests.ts]: remove comment

File: test/ui/viewJobRoleTests.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        await driver.get(url);
+        const title = await driver.findElement(webdriver.By.xpath('/html[1]/body[1]/a[1]/h1[1]')).getText();
+
+        expect(title).to.equal('Open Job Roles');
+    });
+
+    it('Job role table should populate', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://2qr8mnb3c3.eu-west-1.awsapprunner.com/openJobRoles';
+        await driver.get(url);
+        const data = await driver.findElement(webdriver.By.xpath('/html[1]/body[1]/a[1]/table[1]/tbody[1]/tr[2]/td[1]')).getText();
```
Comment (boydstepkainos) [inline test/ui/viewJobRoleTests.ts]: This is not a good selector because it relies on the page not changing, let's use an id

Comment (boydstepkainos) [reply test/ui/viewJobRoleTests.ts]: do this for all selectors

File: test/ui/viewJobRoleTests.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    after(async function () {
+        try {
+            await driver.quit();
+        } catch (error) {
+            console.error('Error quitting the driver:', error);
+        }
+    });
+
+    it('Page should load and title text should display', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://2qr8mnb3c3.eu-west-1.awsapprunner.com/openJobRoles';
```
Comment (boydstepkainos) [inline test/ui/viewJobRoleTests.ts]: this can go in the before, it's the same for all tests

File: test/ui/viewJobRoleTests.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        expect(data).to.not.equal(null);
+    });
+
+    it('Location data should only show valid locations', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://2qr8mnb3c3.eu-west-1.awsapprunner.com/openJobRoles';
+        await driver.get(url);
+
+        const names: string[] = ["BELFAST", "LONDON", "TORONTO", "BIRMINGHAM", "GDANSK"];
+
+        const table = await driver.findElement(webdriver.By.css('table'));
```
Comment (boydstepkainos) [inline test/ui/viewJobRoleTests.ts]: Ah, unfortunately, this isn't reliable either. Anything that relies on page structure could break with further requirements. Please look into adding the ID's like we chatted about and using them as your selectors. It should be less work than all this code :D

File: test/ui/viewJobRoleTests.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import { Builder, WebDriver } from 'selenium-webdriver';
+import webdriver from 'selenium-webdriver';
+import { expect } from 'chai';
+import * as chrome from 'selenium-webdriver/chrome';
+
+describe('Job Roles Test', function () {
+
+    let driver: WebDriver;
+
+    before(async function () {
```
Comment (JasonWKainos) [inline test/ui/viewJobRoleTests.ts:10]: Can we create a seperate class used for initialising the driver. This should be held in a separate class following the Page Object Model setup

File: test/ui/viewJobRoleTests.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    it('Page should load and title text should display', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://2qr8mnb3c3.eu-west-1.awsapprunner.com/openJobRoles';
+        await driver.get(url);
+        const title = await driver.findElement(webdriver.By.css('h1')).getText();
+
+        expect(title).to.equal('Open Job Roles');
+    });
+
+    it('Job role table should populate', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://2qr8mnb3c3.eu-west-1.awsapprunner.com/openJobRoles';
```
Comment (JasonWKainos) [inline test/ui/viewJobRoleTests.ts]: Add another separate class for all of your constants and call them into your test. URLs should not be hardcoded into your actual tests

File: test/ui/viewJobRoleTests.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    it('Date fields should only hold data in a correct date format', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://2qr8mnb3c3.eu-west-1.awsapprunner.com/openJobRoles';
+        await driver.get(url);
+
+        const table = await driver.findElement(webdriver.By.css('table'));
+
+        const rows = await table.findElements(webdriver.By.css('tr'));
+
+        //Regular expression to validate the format of the dates in the closing date field
```
Comment (JasonWKainos) [inline test/ui/viewJobRoleTests.ts]: Create a page class called ViewJobRoles to have these functions. Then you can have a seperate class called ViewJobRolesTest which calls these functions. Your tests should just be a list of calling functions from your relevant page classes.

File: .pa11yci [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
     "urls": [
 
-
+
```
Comment (boydstepkainos) [inline .pa11yci:4]: can undo

File: test/ui/driverBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { Builder, WebDriver } from 'selenium-webdriver';
+import 'chromedriver';
+import * as chrome from 'selenium-webdriver/chrome';
```
Comment (boydstepkainos) [inline test/ui/driverBuilder.ts:2]: only import what you need

File: test/ui/driverBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+export class driverBuilder {
+
+    static driver: WebDriver;
+
+    static async before(): Promise<WebDriver> {
```
Comment (boydstepkainos) [inline test/ui/driverBuilder.ts]: before/after are test concepts (ran before and after each test) - can these functions be renamed?

File: test/ui/viewJobRoleTests.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import 'mocha';
+import { viewJobRolePage } from './viewJobRolePage'; 
+import { driverBuilder } from './driverBuilder';
+
+describe('Job Roles Test', function () {
+    let driver: WebDriver;
+    let jobRolePage: viewJobRolePage;
+
+    before(async function () {
+        driver = await driverBuilder.before();
```
Comment (boydstepkainos) [inline test/ui/viewJobRoleTests.ts]: this is unfortunately still creating a new driver before each test - @JasonWKainos  is that necessary?

**Review by dylanlaffin — APPROVED** [review]

**Review by boydstepkainos — APPROVED** [review]

**Pull request merged**